docs/proposals/apiserver-ha.md

```diff
@@ -0,0 +1,152 @@
+with replicated API servers easily(see
+[high-availability](../../docs/admin/high-availability.md) ).
+However, there are still some problems
+need to address. This proposal aims at address those problems.
+
+### Design
+
+```
+  |--------|            |--------|            |--------|              
+  | etcd-1 |            | etcd-2 |            | etcd-3 |        
+  |--------|            |--------|            |--------|
+      |                     |                     |
+      |                     |                     |
+|-------------|      |-------------|        |-------------|             
+| apiserver-1 |      | apiserver-2 |        | apiserver-3 |         
+|-------------|      |-------------|        |-------------|   
+      |                     |                     |
+      |                     |                     |
+|---------------------------------------------------------|             
+|  Nginx/HAProxy, or client randomly select an endpoint   |         
+|---------------------------------------------------------|
+       |                     |                     |
+       |                     |                     |
+ |------------|       |------------|        |-------------|              
+ |  scheduler |       | controller |        |    node     |      
+ |------------|       |------------|        |-------------|          
+```
+
+### Initional Implementation step
+
+#### 1. Introduce a knob to enable quorum read
+
+There are two cases where we may read old data if we have multi API server:
+
+1) When a etcd member is isolated from etcd cluster, we can still read from
+it(but all the write request sent to the isolated member will failed). If
+unfortunately nginx forward some read request to the isolated member,
+those read request will get old data.
+
+2) Consider the following case:
+
+A write request(e.g. ?key=hello&value=world) was forward to apiserver-1
+and then forward to etcd-1, etcd-1 will return 200 indicating the write
+request was succeeded if the data has been written to a majority of members.
+(for example, written to etcd-1 and etcd-2) Almost immediately, a read
+request(?key=hello) was forward to etcd-3, if API server doesn't enable quorum
+read, this read request will failed(key not found) since the data has not been
+synced to etcd-3.
+
+So, we must introduce a knob to enable quorum read if we have multi API server.
+Since quorum read will make read request less effective, user could disenable it
+if they just want a single API server.
+
+
+#### 2. Default kubernetes service/endpoints
+
+When API server start, it will create a kubernetes service, and will periodically
+reconcile it. Currently, API server assume there is only one endpoint(benckend)
+for the service, every apiserver will constantly overwrites the etcd key
+“/registry/services/endpoints/default/kubernetes". We'd fix this by writing
+multiple endpoints in there.
```

Comment:
No-- there's already a knob--how many apiservers are in the group-- and the endpoints record is arranged such that it will eventually be correct. I don't think any code changes should be necessary.

We do need to customize the service name to support multiple groups of apiservers, but that's unrelated to HA.